docs/proposals/17-ratelimits-for-shoot-dns-entries.md

```diff
@@ -0,0 +1,112 @@
+- Updates of `DNSEntries` created by Gardener (e.g. `internal`, `external`, and `ingress`) should not be blocked by the rate limits.
+- If creation/updates of `DNSEntries` of the `shoot-dns-service` are pending because of a rate limit, this should also made be visible in the shoot cluster (either by events or status update)
+
+### Non-Goals
+
+- prioritize updates between `DNSEntries` assigned to the same `DNSProvider`
+
+## Proposal
+
+This proposal involves changes in two components:
+
+- [dns-controller-manager](https://github.com/gardener/external-dns-management) for implementation of rate limiting
+
+- [gardener](https://github.com/gardener/gardener) for configuration of rate limits, applying them to the external `DNSProvider` and annotating `DNSEntries` to exclude then from rate limiting
+
+### DNS controller manager
+
+The spec of a `DNSProvider` is extended by an optional `rateLimit` section.
+
+```go
+type DNSProviderSpec struct {
+    ...
+    // rate limit for create/update operations on DNSEntries assigned to this provider
+    // +optional
+    RateLimit *RateLimit `json:"rateLimit,omitempty"`
+}
+
+type RateLimit struct {
+    // RequestsPerDay is create/update request rate per DNS entry given by requests per day
+    RequestsPerDay int `json:"requestsPerDay"`
+    // Burst allows bursts of up to 'burst' to exceed the rate defined by 'RequestsPerDay', while still maintaining a
+    // smoothed rate of 'RequestsPerDay'
+    Burst int `json:"burst"`
+}
+```
+
+Example manifest:
+
+```yaml
+kind: DNSProvider
+...
+spec:
+  ...
+  rateLimit:
+    requestsPerDay: 120
+    burst: 20
+```
+
+Each valid `DNSEntry` is already automatically assigned to an `DNSProvider` by matching the domain name and the realm.
+This assignment now also includes the regime of the rate limit of this `DNSProvider`.
+The creation or update of DNS records is postponed until the rate limit restrictions are fulfilled.
+The entry stays in the state `Pending` or `Updating` until the request is applied on the DNS cloud service.
+Here the state `Updating` is a new, additional state to be implemented to make transparent that spec and backend are not yet synchronized.
+
+Deletion of `DNSEntry` is always applied immediately to avoid possible extreme delays on shoot deletion.
+
+As the status of `DNSEntries` created by the `shoot-dns-service` in the CP is already forwarded to the source objects in the shoot cluster in form
+of events, nothing needs to be added here. Additionally, the status for shoot-side `DNSEntries` is updated accordingly.
+
+On replicating a `DNSEntry` from the shoot cluster nothing must be changed, as annotations are not copied. This means a user cannot escape the rate limiting by adding an annotation like `dns.gardener.cloud/not-rate-limited: "true"`
+
+### Gardener
+
+The rate limit for the default external `DNSProvider` is configured for the default domains of the garden.
```

Comment:
Which will probably be another "well known" annotation (e.g. `dns.gardener.cloud/rate-limit`) in the [default domain secret](https://github.com/gardener/gardener/blob/7616e05b6dedd94b4c6e48c26ac026da4d177191/example/10-secret-default-domain.yaml#L13). I thought about alternatives like configuring default values via `ControllerRegistration` or adding support to specify an entire `DNSProvider` in Gardener (via `Unstructured`), but yet another annotation might be the best option to choose for now. WDYT?

Reply:
Can you add this explanation to the GEP?

Reply:
done, see last commit https://github.com/gardener/gardener/pull/4000/commits/9f61c6d558f328bf0cf96fbacb3fd7fa62916db4